keps/sig-cli/20190805-pod-troubleshooting.md

```diff
@@ -0,0 +1,380 @@
+   the script issues docker commands directly and so isn't portable to other
+   runtimes.
+2. The script requires broad administrative access to the node to run docker
+   commands.
+3. Installing tools into the running container requires modifying the image
+   being debugged.
+
+`kubectl debug` would replace the scratch debugger script with a method
+idiomatic to Kubernetes.
+
+### Operations and Support
+
+As Kubernetes gains in popularity, it's becoming the case that a person
+troubleshooting an application is not necessarily the person who built it.
+Operations staff and Support organizations want the ability to attach a "known
+good" or automated debugging environment to a pod.
+
+### Goals
+
+Make available to users of Kubernetes a troubleshooting facility that:
+
+1. works out of the box as a first-class feature of the platform.
+2. does not depend on prior inclusion of tools in container images.
+3. does not require administrative access to the node.
+
+New `kubectl` concepts increase cognitive burden for all users of Kubernetes.
+This KEP seeks to minimize this burden by mirroring the existing `kubectl exec`
+workflow.
+
+### Non-Goals
+
+Ephemeral containers are supported on Windows, but it's not the recommended
+debugging facility. Debugging relies on process namespace sharing, which isn't
+implemented by the Windows container runtime, and Windows containers cannot be
+built `FROM scratch` so debugging tools are already available in the container.
+While not excluding Windows containers, this KEP will focus on the Linux use
+case.
+
+[distroless container image]: https://github.com/GoogleCloudPlatform/distroless
+[distroless for k8s system images]: https://git.k8s.io/enhancements/keps/sig-release/20190316-rebase-images-to-distroless.md
+[scratch debugger]: https://github.com/kubernetes-retired/contrib/tree/master/scratch-debugger
+
+## Proposal
+
+We will add a new command `kubectl debug` that will:
+
+1. Construct a `v1.Container` for the debug container based on command line
+   arguments. This will include an optional container for namespace targeting
+   (described below).
+2. Fetch the specified pod's existing ephemeral containers using
+   `GetEphemeralContainers()` in the generated pod client.
+3. Append the new debug container to the pod's ephemeral containers and call
+   `UpdateEphemeralContainers()`.
+4. [optional] Watch pod for updates to the debug container's `ContainerStatus`
+   and automatically attach once the container is running.
+
+### Debug Container Naming
+
+Currently, there is no support for deleting or recreating ephemeral containers.
+In cases where the user does not specify a name, `kubectl` will generate a
+unique name and display it to the user.
+
+### Container Namespace Targeting
+
+In order to see processes running in other containers in the pod, [Process
+Namespace Sharing] should be enabled for the pod. In cases where process
```

Comment:
This should be called out both in command help and also in constraints section, w/o it `kubectl debug` won't be able to that much.

Reply:
I'll make sure it's called out in command help (https://github.com/verb/kubectl-debug/commit/64404a97ed8d51c9a1b427f4b8a624630120673f, 1a39bf53b1a698232c3ca113cca3fed82e893a4f) but I'm not sure it's a constraint since we will support targeting container namespaces. Based on feedback, here's how I think it will be used:

shareProcessNamespace | container targeted? | See other processes? | Use case
---------------------------------- | ------------------------ | ----------------------------- | -------------
true | - | yes | application troubleshooting
false | yes | yes | application troubleshooting
false | no | no | network troubleshooting

Concrete example for network troubleshooting: kubedns wants to use distroless containers but wants to be able to run an ephemeral container to be able to `dig` against localhost.